blog/2023-09-25-constructs-methods.md

```diff
@@ -0,0 +1,247 @@
+
+```js
+let flower = new Flower(garden, `tulip${i}`, {
+  kind: "tulip",
+  color: "yellow",
+});
+```
+
+But as we saw in the introduction, it's also possible to for methods to change a construct's properties.
+For example, we could add a method that changes the flower's color:
+
+```js
+flower.setColor("blue");
+```
+
+This works like you'd imagine - and it's easy to implement.
+However, it's not without drawbacks.
+
+By making the construct's state *mutable*, it's possible for it to be changed in more than one place. can lead to surprising behavior.
+
+For example, take the following code where I've defined two new constructs, an `OrangePatch` and `PurplePatch`, both accepting a flower in its `props`:
+
+```js
+class OrangePatch extends Construct {
+  constructor(scope, id, props) {
+    super(scope, id, props);
+    props.flower.setColor("orange");
+  }
+}
+
+class PurplePatch extends Construct {
+  constructor(scope, id, props) {
+    super(scope, id, props);
+    props.flower.setColor("purple");
+  }
+}
+
+const garden = new Garden();
+const rose = new Flower(garden, "rose", {
+  kind: "rose",
+  color: "red",
+});
+new OrangePatch(garden, "orange-patch", { flower: rose });
+new PurplePatch(garden, "purple-patch", { flower: rose });
+```
+
+Since they both set the color of `rose`, one of them is going to override the decision of the other (in this case, the final rose will be purple).
+Uh oh!
+
+To avoid these kinds of issues, we usually follow these two rules when designing the methods of constructs.
+
+### Rule 1: Only add state, never subtract or update
+
+Methods should add state, not update or subtract state. If you're always adding state, then state that was configured or added earlier in the application won't get removed or overridden. The additions should also be [commutative](https://en.wikipedia.org/wiki/Commutative_property) - which means re-ordering them should not change the application's functional behavior.
+
+We can see an example of this rule with the `addEnvironment` method on `cloud.Function`:
+
+```js
+let fn = new cloud.Function(/* props */);
+fn.addEnvironment("DB_NAME", "orders");
+```
+
+If you try calling `addEnvironment` with the same string twice, it throws an error.
+Since environment variables can only be added, you can pass around `fn` throughout your codebase - including to third party libraries! - without worrying about environment variables being removed or changed.
+
+Besides changing internal state on a construct, methods can also create new auxiliary constructs, which is another way to safely add state. For example, in Wing it's possible to add objects to a bucket during deployment like so:
+
+```js
+bring cloud;
+
+let bucket = new cloud.Bucket();
+bucket.addObject("hello.txt", "cool beans");
+```
+
+Under the hood, `addObject` creates a new construct representing a bucket object, and that gets translated into IaC configuration separately from the actual bucket.
```

Comment:
I am not sure that's actually different from `addEnvironment()`. The fact that under the hood it creates another object shouldn't matter from an API design perspective.

Reply:
Agree, maybe this tangent is unneeded